rpc/websockets.go

```diff
@@ -0,0 +1,546 @@
+package rpc
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"net"
+	"net/http"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/gorilla/mux"
+	"github.com/gorilla/websocket"
+	"github.com/spf13/viper"
+
+	"github.com/tendermint/tendermint/libs/log"
+	coretypes "github.com/tendermint/tendermint/rpc/core/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+
+	evmtypes "github.com/cosmos/ethermint/x/evm/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/eth/filters"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	context "github.com/cosmos/cosmos-sdk/client/context"
+)
+
+// SubscriptionResponseJSON for json subscription responses
+type SubscriptionResponseJSON struct {
+	Jsonrpc string      `json:"jsonrpc"`
+	Result  interface{} `json:"result"`
+	ID      float64     `json:"id"`
+}
+
+type SubscriptionNotification struct {
+	Jsonrpc string              `json:"jsonrpc"`
+	Method  string              `json:"method"`
+	Params  *SubscriptionResult `json:"params"`
+}
+
+type SubscriptionResult struct {
+	Subscription rpc.ID      `json:"subscription"`
+	Result       interface{} `json:"result"`
+}
+
+// ErrorResponseJSON json for error responses
+type ErrorResponseJSON struct {
+	Jsonrpc string            `json:"jsonrpc"`
+	Error   *ErrorMessageJSON `json:"error"`
+	ID      *big.Int          `json:"id"`
+}
+
+// ErrorMessageJSON json for error messages
+type ErrorMessageJSON struct {
+	Code    *big.Int `json:"code"`
+	Message string   `json:"message"`
+}
+
+type websocketsServer struct {
+	rpcAddr string // listen address of rest-server
+	wsAddr  string // listen address of ws server
+	api     *pubSubAPI
+	logger  log.Logger
+}
+
+func newWebsocketsServer(cliCtx context.CLIContext, wsAddr string) *websocketsServer {
+	return &websocketsServer{
+		rpcAddr: viper.GetString("laddr"),
+		wsAddr:  wsAddr,
+		api:     newPubSubAPI(cliCtx),
+		logger:  log.NewTMLogger(log.NewSyncWriter(os.Stdout)).With("module", "websocket-server"),
+	}
+}
+
+func (s *websocketsServer) start() {
+	ws := mux.NewRouter()
+	ws.Handle("/", s)
+
+	go func() {
+		err := http.ListenAndServe(fmt.Sprintf(":%s", s.wsAddr), ws)
+		if err != nil {
+			s.logger.Error("http error:", err)
+		}
+	}()
+}
+
+func (s *websocketsServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var upgrader = websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool {
```

Comment:
we should consider updating it (maybe for a follow-up PR): https://github.com/ethereum/go-ethereum/blob/master/rpc/websocket.go#L69